File: library/src/main/java/com/benny/library/neobinding/bind/OneWayPropertyBinding.java
package com.benny.library.neobinding.bind;

import android.util.Log;
import com.benny.library.neobinding.converter.EmptyConverter;
import com.benny.library.neobinding.converter.MultiplePropertyConverter;
import com.benny.library.neobinding.converter.OneWayPropertyConverter;
import org.apache.commons.lang3.StringUtils;
import rx.Observable;
import rx.functions.Action1;
import rx.functions.Func1;
import rx.subjects.Subject;

import java.util.List;

/**
 * Created by benny on 11/5/15.
 */
public class OneWayPropertyBinding<T> {
    private String property;
    private List<String> properties;

    private OneWayPropertyConverter<?> converter;
    private MultiplePropertyConverter<T> multipleConverter;

    private Action1<? super T> observer;
    private Observable<T> observable;

    public OneWayPropertyBinding(String property, Action1<T> observer) {
        this.property = property;
        this.converter = new EmptyConverter<T>();
        this.observer = observer;
    }

    public OneWayPropertyBinding(String property, OneWayPropertyConverter<T> converter, Action1<? super T> observer) {
        this.property = property;
        this.converter = converter;
        this.observer = observer;
    }

    public OneWayPropertyBinding(String property, Observable<T> observable) {
        this.property = property;
        this.converter = new EmptyConverter<T>();
        this.observable = observable;
    }

    public OneWayPropertyBinding(String property, OneWayPropertyConverter<?> converter, Observable<T> observable) {
        this.property = property;
        this.converter = converter;
        this.observable = observable;
    }

    public OneWayPropertyBinding(List<String> properties, MultiplePropertyConverter<T> multipleConverter, Action1<? super T> observer) {
        this.properties = properties;
        this.multipleConverter = multipleConverter;
        this.observer = observer;
    }

    public String property() {
        return property != null ? property : properties.get(0);
    }

    public List<String> properties() {
        return properties;
    }

    public void bindTo(BindingContext<?> bindingContext, List<Observable<?> > observables) {
        Observable.combineLatest(observables, multipleConverter::convert).compose(bindingContext.applyLifecycle())
                .doOnSubscribe(() -> ViewModelBinder.LogBinding(StringUtils.join(properties, ",") + " multiple"))
                .doOnTerminate(() -> ViewModelBinder.LogUnbinding(StringUtils.join(properties, ",") + " multiple"))
                .doOnNext(o-> Log.d("Binding", "multiple onNext:" + o))
                        .subscribe(observer);
    }

    @SuppressWarnings("unchecked")
    public void bindTo(BindingContext<?> bindingContext, Subject subject) {
        if(observable != null) {
            observable.map((Func1<T, Object>)converter::convert).compose(bindingContext.applyLifecycle())
                    .doOnSubscribe(() -> ViewModelBinder.LogBinding(property + " view->model"))
                    .doOnTerminate(() -> ViewModelBinder.LogUnbinding(property + " view->model"))
                    .subscribe(subject);
        }
        else {
            subject.map(converter::convert).compose(bindingContext.applyLifecycle())
                    .doOnSubscribe(() -> ViewModelBinder.LogBinding(property + " model->view"))
                    .doOnTerminate(() -> ViewModelBinder.LogUnbinding(property + " model->view"))
                    .subscribe(observer);
        }
    }
}

File: library/src/main/kotlin/com/benny/library/neobinding/bind/OneWayPropertyBinding.kt
package com.benny.library.neobinding.bind

import com.benny.library.neobinding.converter.EmptyOneWayConverter
import com.benny.library.neobinding.converter.MultipleConverter
import com.benny.library.neobinding.converter.OneWayConverter
import rx.Observable
import rx.functions.Action1

/**
 * Created by benny on 11/17/15.
 */

public class OneWayPropertyBinding<T, R> private constructor(key: String) : PropertyBinding() {
    public var key:String = key
        private set

    var converter: OneWayConverter<R>? = null
    var backConverter: OneWayConverter<T>? = null

    var observable: Observable<T>? = null
    var observer: Action1<in T>? = null

    constructor(key: String, observable: Observable<T>, converter: OneWayConverter<R>? = EmptyOneWayConverter<R>()) : this(key) {
        this.observable = observable
        this.converter = converter ?: EmptyOneWayConverter<R>()
    }

    constructor(key: String, observer: Action1<in T>, backConverter: OneWayConverter<T>? = EmptyOneWayConverter<T>()) : this(key) {
        this.observer = observer
        this.backConverter = backConverter ?: EmptyOneWayConverter<T>()
    }

    public fun prefix(prefix: String) : OneWayPropertyBinding<T, R> {
        if(!prefix.isEmpty()) key = "$prefix.$key"
        return this
    }

    public fun bindTo(bindingContext: BindingContext, property: Property<R>) {
        if (observable != null) {
            observable!!.compose(bindingContext.applyLifecycle<T>())
                    .map { (converter as OneWayConverter<R>).convert(it as Any) }
                    .subscribe(property.observer)
        } else {
            property.observable
                    .compose(bindingContext.applyLifecycle<R>())
                    .map { backConverter!!.convert(it as Any) }
                    .subscribe(observer)
        }
    }

    public fun bindTo(property: Property<R>) {
        if (observable != null) {
            observable!!.map { (converter as OneWayConverter<R>).convert(it as Any) }
                    .subscribe(property.observer)
        } else {
            property.observable.map { backConverter!!.convert(it as Any) }
                    .subscribe(observer)
        }
    }
}